guide.md

```diff
@@ -0,0 +1,350 @@
+
+```bazel
+load("@io_tweag_rules_nixpkgs//nixpkgs:nixpkgs.bzl", "nixpkgs_cc_configure")
+nixpkgs_cc_configure(
+  repository = "@nixpkgs",
+  name = "nixpkgs_config_cc",
+)
+```
+
+This will select the default `stdenv.cc` package as the compiler, drawing
+auxiliary tools from `stdenv.cc.bintools`. However, one can write a little bit of
+Nix to configure this:
+
+```bazel
+load("@io_tweag_rules_nixpkgs//nixpkgs:nixpkgs.bzl", "nixpkgs_cc_configure")
+nixpkgs_cc_configure(
+  repository = "@nixpkgs",
+  # nix_file_content = "(import <nixpkgs> {}).gcc11",
+  attribute_path = "gcc11",
+  name = "nixpkgs_config_cc",
+)
+```
+
+`(import <nixpkgs> {})` loads the entire package set into an attribute set
+(Nix's equivalent of a record), and `.gcc11` selects one value from it. It is
+possible to customize the compiler further using more Nix if desired: you could,
+for instance, point Nix to a custom fork of Clang containing support for a new
+architecture, and have everything work seamlessly.
+
+With this, we now have a fully hermetic CC toolchain that will behave correctly
+across platforms. It remains to tell Bazel to use it by default, instead of
+`@local_config_cc`. To do so, we put the following into `.bazelrc` at the root
+of the repository:
+
+```
+build --host_platform=@io_tweag_rules_nixpkgs//nixpkgs/platforms:host
+build --crosstool_top=@nixpkgs_config_cc//:toolchain
+```
+
+Now `bazel build` and `bazel run` should use this toolchain by default for
+C/C++ compilation:
+```
+> bazel build //src:hello-world
+INFO: Analyzed target //src:hello-world (42 packages loaded, 177 targets configured).
+INFO: Found 1 target...
+Target //src:hello-world up-to-date:
+  bazel-bin/src/hello-world
+INFO: Elapsed time: 0.377s, Critical Path: 0.22s
+INFO: 6 processes: 4 internal, 2 linux-sandbox.
+INFO: Build completed successfully, 6 total actions
+```
+
+# Making Nix optional
+
+It is entirely possible to make the use of Nix optional in a Bazel setup. To do
+so, we use `bazelrc` configs to only enable Nix-specific configuration when a
+`--config` flag is provided, like so:
+
+```
+build:nix --host_platform=@io_tweag_rules_nixpkgs//nixpkgs/platforms:host
+build:nix --crosstool_top=@nixpkgs_config_cc//:toolchain
+```
+
+None of these options will apply to `bazel build` or `bazel run` unless an
+extra `--config nix` parameter is passed:
+
+```
+> bazel run //src:hello-world --config=nix
+...
+```
+
+# Working from a Nix-provided developer environment
+
+The `nix-shell` command we used to get a Nix-provided Bazel was the simplest
+example of a Nix developer environment, but they can provide developers on a
+complex project with a lot more convenience. We can create development
+environments that all contributors on a project can use with zero setup. These
+shells provide consistent versions of libraries for whichever languages are
+used in the project as well as for system dependencies.
+
+We haven't _disabled_ Bazel's deep-rooted tendency to depend on the global
```

Comment:
This is not entirely true, or at least confusing for me. One could argue that we partially disabled it earlier when we used

```
build --host_platform=@io_tweag_rules_nixpkgs//nixpkgs/platforms:host
build --crosstool_top=@nixpkgs_config_cc//:toolchain
```

Maybe rephrase it to make clear that without special care and options like the ones above Bazel might be influenced by its environment?

Reply:
Done.